### ISSUE TITLE
unsafe eval is not an allowed source error (CSP not being modified for disqus.com iframes)

### ISSUE THREAD

**Issue by trlkly** [Tampermonkey/tampermonkey#776]
## Expected Behavior

Scripts that inject into the unsafe window will run properly. The Content Security Policy of the page will be altered to allow injection from Tampermonkey scripts, per the Tampermonkey settings. 

## Actual Behavior

Scripts are not allowed to run, generating the following error in the error console:
> Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src https:".

## Specifications

- Chromium:  77.0.3865.120
- TM: 4.8.41
- OS: Windows 7 SP1

## Script

```javascript
// ==UserScript==
// @name         scratchpad
// @include      *
// @grant        none
// ==/UserScript==

window.addCSS = function (css) {
  var style = document.createElement('style');
  style.textContent = css;
  document.documentElement.appendChild(style);
  return style;
};
```
## Example sites with problems:

* http://namesakecomic.com/comic/time-to-work
* http://www.aliceandthenightmare.com/comic/patron-questions-5
* http://www.gogetaroomie.com/comic/seen-worse

(Commonality appears to be including an embedded disqus.com comment section.)

**Comment by trlkly** [comment]
Upon testing now, this error even shows up when I only activate a single script, and leave it blank other than the UserScript directive. (I did this to make sure I was not in any way touching the window object).

I've also noticed that the scripts do actually work outside of the iframe from disqus.com. The issue may be that iframes do not get their CSP changed.

**Comment by trlkly** [comment]
In case it is relevant, I am 100% sure it's a CSP issue, as I was able to use a CSP modifying extension to workaround the problem. However, it is my understanding that this is also supposed to be built into TamperMonkey as well.

Here is the extension I used: [Content Security Policy Override](https://chrome.google.com/webstore/detail/content-security-policy-o/lhieoncdgamiiogcllfmboilhgoknmpi). 

And all I did was add "unsafe-eval" to the script-src tag. I didn't know anything more specific to TamperMonkey to try.

**Comment by derjanb** [comment]
This is working fine here. Did you maybe modify the setting called "Add Tampermonkey to the site's content security policy (CSP) if there is one"? Are there other extension which try to modify the request?

**Comment by trlkly** [comment]
No, that setting is set properly. I even toggled it both on and off to see if the setting needed to be reset. I also turned off all extensions other than TamperMonkey, and the problem still occurred on the links given. 

Here are my Tampermonkey settings, if they are relevant:

> {"created_by":"Tampermonkey","version":"1","settings":{"action_menu_columns":"2","action_menu_scripts_hide_disabled":false,"action_menu_scripts_sort":"position","appearance_badge_color":"#ee3131","appearance_badges":"off","configMode":100,"connect_mode":"ask","editor_autoLint":true,"editor_autoLintMaxLen":"50000","editor_autoSave":false,"editor_easySave":true,"editor_electricChars":true,"editor_enabled":true,"editor_fontSize":"120","editor_highlightTrailingWhitespace":true,"editor_indentUnit":"2","editor_indentWithTabs":"spaces","editor_keyMap":"windows","editor_lineWrapping":false,"editor_linter_config":null,"editor_tabMode":"classic","editor_theme":"default","editor_trimTrailingSpacesFromModifiedLines":true,"enabled":true,"external_connect":"all","favicon_service":"google","forbiddenPages":["*example.org/*","*paypal.tld/*","*stripe.com/*","https://*deutsche-bank-24.tld/*","https://*bankofamerica.tld/*","/^.*:\\/\\/apis\\.google\\.com\\/((?!render)([^\\/]+)\\/)+([^\\/]+)?$/","*://www.facebook.com/plugins/*","*://platform.twitter.com/widgets/*"],"incognito_mode":"temporary","layout":"default","layout_user_css":".head_container { display: none }\n.tv_container_fit { top: 0 }\n.nameNicon64>img { display: none; width: 1em; height 1em }\n.editorbuttonbar { display: initial; }\n#action { font-size: 75%; }\n.script_name { max-width: 15em !important; }","native_import":false,"native_import_path":"H:\\Users\\[NAME_REDACTED]\\AppData\\Local\\Google\\Chrome\\User Data\\Profile 1\\Extensions\\dhdgffkkebhmkfjojejmpbldmpobfkfo","native_import_post_action":"uninstall","notification_showUpdate":"changelog","page_filter_mode":"black","page_whitelist":["/https?:\\/\\/greasyfork\\.org\\/.*/","http://xkcd.com/970/"],"require_sri_mode":"supported","reset_sync":null,"runtime_inject_mode":"instant","scriptUrlDetection":"auto","script_blacklist_server":[{"rules":["/https?:\\/\\/userscripts\\.org\\/scripts\\/(source|version)/(102614).*\\.user\\.js/"],"severity":"2"},{"name":"SaveFrom.net helper","reason":"Shows advertisements","rules":["/https?:\\/\\/sf-addon\\.com\\/.*\\/.*\\.user\\.js/","/https?:\\/\\/.*\\.sf-helper\\.com\\/.*\\.user\\.js/"],"severity":"3"}],"script_file_access":"externals","script_templates":[{"name":"ECMAScript 5","value":"// ==UserScript==\n// @name         New Userscript\n// @namespace    http://tampermonkey.net/\n// @version      0.1\n// @description  try to take over the world!\n// @author       You\n// @match        <$URL$>\n// @grant        none\n// ==/UserScript==\n\n(function() {\n    'use strict';\n\n    // Your code here...\n})();"},{"name":"ECMAScript 6","value":"// ==UserScript==\n// @name         New ES6-Userscript\n// @namespace    http://tampermonkey.net/\n// @version      0.1\n// @description  shows how to use babel compiler\n// @author       You\n// @require      https://cdnjs.cloudflare.com/ajax/libs/babel-standalone/6.18.2/babel.js\n// @require      https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/6.16.0/polyfill.js\n// @match        <$URL$>\n// ==/UserScript==\n\n/* jshint ignore:start */\nvar inline_src = (<><![CDATA[\n/* jshint ignore:end */\n    /* jshint esnext: false */\n    /* jshint esversion: 6 */\n\n    // Your code here...\n\n/* jshint ignore:start */\n]]></>).toString();\nvar c = Babel.transform(inline_src, { presets: [ \"es2015\", \"es2016\" ] });\neval(c.code);\n/* jshint ignore:end */"},{"name":"CoffeeScript","value":"// ==UserScript==\n// @name         New Coffee-Userscript\n// @namespace    http://tampermonkey.net/\n// @version      0.1\n// @description  shows how to use coffeescript compiler\n// @author       You\n// @require      http://coffeescript.org/extras/coffee-script.js\n// @match        <$URL$>\n// ==/UserScript==\n/* jshint ignore:start */\nvar inline_src = (<><![CDATA[\n\n    // Your code here\n\n]]></>).toString();\nvar compiled = this.CoffeeScript.compile(inline_src);\neval(compiled);\n/* jshint ignore:end */"}],"sync_enabled":true,"sync_type":"2","webrequest_fixCSP":"yes","webrequest_modHeaders":"yes"}}

**Comment by trlkly** [comment]
Now even the extension I was using seems to no longer allow overriding the CSP when a page includes disqus.com iframes.  But I get no errors when loading disqus.com directly, even if I turn that extension off. 

The problem seems to be that the iframe CSP is not being modified.

**Comment by trlkly** [comment]
This is utterly bizarre. I finally tried disabling the "Add Tampermonkey to the site's content security policy (CSP) if there is one", and the addons are suddenly working. 

I even checked the HTTP header. The CSP does not include 'unsafe-eval', but things are working! Disabling a feature should not make it work!

**Comment by raszpl** [comment]
https://csp-evaluator.withgoogle.com/ claims disqus doesnt set any CSP headers

**Comment by trlkly** [comment]
The main site doesn't, but the iframe-embeded discussions on third-party
sites all do. I checked it directly using Chrome's Network tab in its
development console.

The weird thing remains that removing 'unsafe-eval' from the CSP (by
disabling the aforementioned option in Tampermonkey and not using any other
CSP-affecting extensions) actually allowed scripts to run again.

**Comment by derjanb** [comment]
FWIW: I found an issue which could break a page's CSP and fixed it in TM BETA 4.11.6113. Besides this all page are working fine here using Tampermonkey 4.10.

Please re-open the issue if you still experience this issue. Thanks for reporting.
